bpo-38314: Add is_reading() method to asyncio _UnixReadPipeTransport

calluw opened the pull request:
Unix read pipe (from `Lib:asyncio.unix_events`) is missing `is_reading()` method available for other read transports within the `asyncio` protocol, which checks that the transport read is not
paused or the transport is not closing. 

This behaviour is documented in `Doc/library/asyncio-protocol.rst`.

Additional tests for the `Lib:asyncio.unix_events` module which check against the state of the pipe and the behaviour of `is_reading()`.

<!-- issue-number: [bpo-38314](https://bugs.python.org/issue38314) -->
https://bugs.python.org/issue38314
<!-- /issue-number -->

Commit:
Add is_reading() method to _UnixReadPipeTransport

Unix read pipe is missing method available for other read transports
within the aynscio protocol, which checks that the transport read is not
paused or the transport is not closing.

Commit:
Add is_reading() method to _UnixReadPipeTransport

Unix read pipe is missing method available for other read transports
within the aynscio protocol, which checks that the transport read is not
paused or the transport is not closing.

commodo committed:
bpo-38684: haslib: fix build when Blake2 not enabled in OpenSSL (#17043)

asvetlov reviewed (changes requested):
Please fix my comments.
The code change itself is correct.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner committed:
bpo-38644: Add Py_EnterRecursiveCall() to the limited API (GH-17046)

Provide Py_EnterRecursiveCall() and Py_LeaveRecursiveCall() as
regular functions for the limited API. Previously, there were defined
as macros, but these macros didn't work with the limited API which
cannot access PyThreadState.recursion_depth field.

Remove _Py_CheckRecursionLimit from the stable ABI.

Add Include/cpython/ceval.h header file.

aeros commented on Misc/NEWS.d/next/Library/2019-11-04-17-20-43.bpo-38314.zWz6_P.rst:
```diff
@@ -0,0 +1 @@
+The read pipe on UNIX transports now has access to the `is_reading()` API which is noted in the documentation for asyncio protocols as being generally available on read transports.
```
I think we could make this more succinct and briefly mention that it's related to asyncio. Also, the max line width of MISC/news entries has to be within 79 chars. 
```suggestion
The read pipe in :mod:`asyncio` UNIX transports now has access to the
`is_reading()` API.
```

calluw committed:
Update Misc/NEWS.d/next/Library/2019-11-04-17-20-43.bpo-38314.zWz6_P.rst

Co-Authored-By: Kyle Stanley <aeros167@gmail.com>

vstinner committed:
bpo-38644: Pass tstate to Py_EnterRecursiveCall() (GH-16997)

* Add _Py_EnterRecursiveCall() and _Py_LeaveRecursiveCall() which
  require a tstate argument.
* Pass tstate to _Py_MakeRecCheck() and  _Py_CheckRecursiveCall().
* Convert Py_EnterRecursiveCall() and Py_LeaveRecursiveCall() macros
  to static inline functions.

_PyThreadState_GET() is the most efficient way to get the tstate, and
so using it with _Py_EnterRecursiveCall() and
_Py_LeaveRecursiveCall() should be a little bit more efficient than
using Py_EnterRecursiveCall() and Py_LeaveRecursiveCall() which use
the "slower" PyThreadState_GET().

vstinner committed:
bpo-38644: Pass tstate to _Py_CheckFunctionResult() (GH-17050)

* Add tstate parameter to _Py_CheckFunctionResult()
* Add _PyErr_FormatFromCauseTstate()
* Replace PyErr_XXX(...) with _PyErr_XXX(state, ...)

micha2718l committed:
Fix a typo in wave module docstring (GH-17009)

s/pathing/patching/

benjaminp committed:
closes bpo-37633: Reëxport some function compatibility wrappers for macros in ``pythonrun.h``. (GH-17056)

vstinner committed:
_json.c: use Py_UNUSED() macro (GH-17053)

Remove UNUSED macro: use Py_UNUSED() macro instead.

Seluj78 committed:
Update interpreter.rst (GH-17059)

Fixed what seemed to be a weird phrasing.

eduardo-elizondo committed:
bpo-35381 Remove all static state from posixmodule (GH-15892)



After #9665, this moves the remaining types in posixmodule to be heap-allocated to make it compatible with PEP384 as well as modifying all the type accessors to fully make the type opaque.

The original PR that got messed up a rebase: https://github.com/python/cpython/pull/10854. All the issues in that commit have now been addressed since https://github.com/python/cpython/pull/11661 got committed.

This change also removes any state from the data segment and onto the module state itself.


https://bugs.python.org/issue35381



Automerge-Triggered-By: @encukou

jdemeyer committed:
bpo-37645: add new function _PyObject_FunctionStr() (GH-14890)



Additional note: the `method_check_args` function in `Objects/descrobject.c` is written in such a way that it applies to all kinds of descriptors. In particular, a future re-implementation of `wrapper_descriptor` could use that code.

CC @vstinner @encukou 


https://bugs.python.org/issue37645



Automerge-Triggered-By: @encukou

ammaraskar committed:
bpo-38696: Fix usage example of HTTPStatus (GH-17066)

benjaminp committed:
bpo-38692: Add os.pidfd_open. (GH-17063)

benjaminp committed:
closes bpo-38713: Expose P_PIDFD in os if it's defined. (GH-17071)



https://bugs.python.org/issue38713

l0rb committed:
bpo-38716: stop rotating handlers from setting inherited namer and rotator to None (GH-17072)

rhettinger committed:
bpo-38382: Document the early-out behavior for a zero (GH-17037)

vsajip committed:
bpo-36876: Moved Parser/listnode.c statics to interpreter state. (GH-16328)

l0rb committed:
update a deprecated assert in logging tests (GH-17079)

vstinner committed:
bpo-38733: PyErr_Occurred() caller must hold the GIL (GH-17080)

bpo-3605, bpo-38733: Optimize _PyErr_Occurred(): remove "tstate ==
NULL" test.

Py_FatalError() no longer calls PyErr_Occurred() if called without
holding the GIL. So PyErr_Occurred() no longer has to support
tstate==NULL case.

_Py_CheckFunctionResult(): use directly _PyErr_Occurred() to avoid
explicit "!= NULL" test.

methane committed:
bpo-38613: Optimize set operations of dict keys. (GH-16961)

corona10 committed:
bpo-22367: Add tests for fcntl.lockf(). (GH-17010)

vstinner committed:
bpo-38644: Add _PyObject_VectorcallTstate() (GH-17052)

* Add _PyObject_VectorcallTstate() function: similar to
  _PyObject_Vectorcall(), but with tstate parameter
* Add tstate parameter to _PyObject_MakeTpCall()

susan-shu-c committed:
Clarify amount of dots between package and subpackage (GH-17092)

serhiy-storchaka committed:
bpo-38635: Simplify decoding the ZIP64 extra field and make it tolerant to extra data. (GH-16988)

rhettinger committed:
Minor readability improvement for argument handling in itertools.repeat()  (GH-17101)

rhettinger committed:
bpo-38761: Register WeakSet as a MutableSet (GH-17104)

brandtbucher committed:
bpo-38438: Simplify argparse "star nargs" usage. (GH-17106)

jonathan-scholbach committed:
bpo-38771:  Explict test for None in code example (GH-17108)

Zheaoli committed:
bpo-38565: add new cache_parameters method for lru_cache (GH-16916)

rhettinger committed:
bpo-38385: Fix iterator/iterable terminology in statistics docs (GH-17111)

Commit:
Add new test case file for UNIX pipe transport functional tests

Commit:
Merge branch 'fix-issue-38314' of github.com:callumquick/cpython into fix-issue-38314
